File: src/inputs/plugins/air_quality/connector/pms5003.py
import asyncio

import serial

from inputs.plugins.air_quality.connector.base import (
    AirQualityConnector,
    AirQualityData,
)


class PMS5003Connector(AirQualityConnector):
    """
    Air quality connector for PMS5003/PMS7003 particulate matter sensor.

    Reads PM1.0, PM2.5, PM10 via UART/Serial.
    Sensor sends 32-byte frames at 9600 baud.

    Wiring:
        VCC  → 5V
        GND  → GND
        TX   → RX (e.g. /dev/ttyUSB0 or /dev/ttyAMA0)

    Datasheet: https://www.aqmd.gov/docs/default-source/aq-spec/resources-page/plantower-pms5003-manual_v2-3.pdf
    """

    BAUD_RATE = 9600
    FRAME_LENGTH = 32
    FRAME_START = (0x42, 0x4D)

    def __init__(self, config: dict):
        """
        Parameters
        ----------
        config : dict
            Must contain:
            - port (str): serial port, e.g. '/dev/ttyUSB0'
            - location (str, optional): location label, default 'Robot'
        """
        super().__init__(config)
        self.port: str = config.get("port", "/dev/ttyUSB0")
        self.location: str = config.get("location", "Robot")
        self._serial: serial.Serial | None = None

    async def connect(self) -> bool:
        """Open the serial port and connect to the PMS5003/7003 sensor."""
        try:
            self._serial = serial.Serial(
                self.port, baudrate=self.BAUD_RATE, timeout=2.0
            )
            self.logger.info(f"PMS5003Connector: connected on {self.port}")
            return True
        except serial.SerialException as e:
            self.logger.error(f"PMS5003Connector: failed to connect: {e}")
            return False

    async def disconnect(self) -> None:
        """Close the serial port if open."""
        if self._serial and self._serial.is_open:
            self._serial.close()
            self.logger.info("PMS5003Connector: disconnected")

    async def read(self) -> AirQualityData | None:
        """
        Read one frame from PMS5003 sensor.

        Returns
        -------
        AirQualityData or None
            Parsed particulate matter data, or None if read failed.
        """
        if self._serial is None or not self._serial.is_open:
            self.logger.error("PMS5003Connector: not connected")
            return None

        try:
            # Run blocking serial read in executor to avoid blocking event loop
            loop = asyncio.get_event_loop()
            frame = await loop.run_in_executor(None, self._read_frame)
            if frame is None:
                return None
            return self._parse(frame)

        except Exception as e:
            self.logger.error(f"PMS5003Connector: read error: {e}")
            return None

    def _read_frame(self) -> bytes | None:
        """
        Read and validate one 32-byte PMS5003 frame (blocking).

        Returns
        -------
        bytes or None
            Raw 32-byte frame, or None if invalid/timeout.
        """
        # Sync to frame start bytes 0x42 0x4D
        if self._serial is None:
            return None
        while True:
            byte = self._serial.read(1)
            if not byte:
                return None
            if byte[0] == self.FRAME_START[0]:
                next_byte = self._serial.read(1)
                if next_byte and next_byte[0] == self.FRAME_START[1]:
                    break

        rest = self._serial.read(self.FRAME_LENGTH - 2)
        if len(rest) != self.FRAME_LENGTH - 2:
            return None

        frame = bytes(self.FRAME_START) + rest

        # Validate checksum — sum of bytes 0..29 == bytes 30..31 (big-endian)
        checksum = sum(frame[:30]) & 0xFFFF
        expected = (frame[30] << 8) | frame[31]
        if checksum != expected:
            self.logger.warning("PMS5003Connector: checksum mismatch")
            return None

        return frame

    def _parse(self, frame: bytes) -> AirQualityData:
        """
        Parse validated 32-byte PMS5003 frame.

        Frame layout (big-endian, each field 2 bytes):
            [4:6]   PM1.0 standard
            [6:8]   PM2.5 standard
            [8:10]  PM10  standard

        Parameters
        ----------
        frame : bytes
            Validated 32-byte frame.

        Returns
        -------
        AirQualityData
        """

        def word(offset: int) -> int:
            return (frame[offset] << 8) | frame[offset + 1]

        pm25 = float(word(6))
        pm10 = float(word(8))

        # Estimate AQI from PM2.5 using US EPA breakpoints
        aqi = self._pm25_to_aqi(pm25)

        return AirQualityData(
            aqi=aqi,
            pm25=pm25,
            pm10=pm10,
            location=self.location,
            source="pms5003",
        )

    @staticmethod
    def _pm25_to_aqi(pm25: float) -> int:
        """
        Convert PM2.5 concentration to AQI using US EPA formula.

        Parameters
        ----------
        pm25 : float
            PM2.5 in µg/m³.

        Returns
        -------
        int
            Estimated AQI value.
        """
        # (C_low, C_high, AQI_low, AQI_high)
        breakpoints = [
            (0.0, 12.0, 0, 50),
            (12.1, 35.4, 51, 100),
            (35.5, 55.4, 101, 150),
            (55.5, 150.4, 151, 200),
            (150.5, 250.4, 201, 300),
            (250.5, 350.4, 301, 400),
            (350.5, 500.4, 401, 500),
        ]
        for c_low, c_high, aqi_low, aqi_high in breakpoints:
            if c_low <= pm25 <= c_high:
                aqi = ((aqi_high - aqi_low) / (c_high - c_low)) * (
                    pm25 - c_low
                ) + aqi_low
                return round(aqi)
        return 500

File: plugins/inputs/air_quality/connector/pms5003/pms5003.go
package pms5003

import (
	"context"
	"log"
	"time"

	"go.bug.st/serial"

	"github.com/openmind/om1/plugins/inputs/air_quality/connector"
)

const (
	baudRate    = 9600
	frameLength = 32
)

var frameStart = [2]byte{0x42, 0x4D}

// Config mirrors the `config: dict` passed into Python's PMS5003Connector.__init__.
type Config struct {
	Port     string // config["port"], default "/dev/ttyUSB0"
	Location string // config["location"], default "Robot"
}

// Connector implements connector.AirQualityConnector for the PMS5003/PMS7003 sensor.
// Mirrors Python's PMS5003Connector.
type Connector struct {
	cfg    Config
	port   serial.Port
	logger *log.Logger
}

// New creates a new PMS5003 connector, applying the same defaults as the Python version.
func New(cfg Config, logger *log.Logger) *Connector {
	if cfg.Port == "" {
		cfg.Port = "/dev/ttyUSB0"
	}
	if cfg.Location == "" {
		cfg.Location = "Robot"
	}
	if logger == nil {
		logger = log.Default()
	}
	return &Connector{cfg: cfg, logger: logger}
}

func (c *Connector) Name() string {
	return "pms5003"
}

// Connect opens the serial port. Mirrors Python's `connect()`.
func (c *Connector) Connect(ctx context.Context) (bool, error) {
	mode := &serial.Mode{
		BaudRate: baudRate,
		DataBits: 8,
		Parity:   serial.NoParity,
		StopBits: serial.OneStopBit,
	}

	port, err := serial.Open(c.cfg.Port, mode)
	if err != nil {
		c.logger.Printf("PMS5003Connector: failed to connect: %v", err)
		return false, nil
	}
	port.SetReadTimeout(2 * time.Second)

	c.port = port
	c.logger.Printf("PMS5003Connector: connected on %s", c.cfg.Port)
	return true, nil
}

// Disconnect closes the serial port if open. Mirrors Python's `disconnect()`.
func (c *Connector) Disconnect(ctx context.Context) error {
	if c.port != nil {
		if err := c.port.Close(); err != nil {
			return err
		}
		c.logger.Println("PMS5003Connector: disconnected")
	}
	return nil
}

// Read reads one frame from the sensor. Runs in a goroutine, mirroring Python's
// `loop.run_in_executor` usage to avoid blocking the caller.
func (c *Connector) Read(ctx context.Context) (*connector.AirQualityData, error) {
	if c.port == nil {
		c.logger.Println("PMS5003Connector: not connected")
		return nil, nil
	}

	type result struct {
		data *connector.AirQualityData
		err  error
	}
	resCh := make(chan result, 1)

	go func() {
		frame, err := c.readFrame()
		if err != nil {
			c.logger.Printf("PMS5003Connector: read error: %v", err)
			resCh <- result{nil, nil}
			return
		}
		if frame == nil {
			resCh <- result{nil, nil}
			return
		}
		resCh <- result{c.parse(frame), nil}
	}()

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case res := <-resCh:
		return res.data, res.err
	}
}

// readFrame reads and validates one 32-byte frame (blocking).
// Mirrors Python's `_read_frame()`.
func (c *Connector) readFrame() ([]byte, error) {
	one := make([]byte, 1)

	// Sync to frame start bytes 0x42 0x4D
	for {
		n, err := c.port.Read(one)
		if err != nil {
			return nil, err
		}
		if n == 0 {
			return nil, nil // timeout, mirrors Python's "if not byte: return None"
		}
		if one[0] == frameStart[0] {
			n2, err := c.port.Read(one)
			if err != nil {
				return nil, err
			}
			if n2 > 0 && one[0] == frameStart[1] {
				break
			}
		}
	}

	rest := make([]byte, frameLength-2)
	total := 0
	for total < len(rest) {
		n, err := c.port.Read(rest[total:])
		if err != nil {
			return nil, err
		}
		if n == 0 {
			break
		}
		total += n
	}
	if total != len(rest) {
		return nil, nil // incomplete frame, mirrors Python's length check -> None
	}

	frame := append([]byte{frameStart[0], frameStart[1]}, rest...)

	// Validate checksum: sum of bytes 0..29 == bytes 30..31 (big-endian)
	var checksum uint32
	for _, b := range frame[:30] {
		checksum += uint32(b)
	}
	checksum &= 0xFFFF
	expected := uint32(frame[30])<<8 | uint32(frame[31])
	if checksum != expected {
		c.logger.Println("PMS5003Connector: checksum mismatch")
		return nil, nil
	}

	return frame, nil
}

// parse parses a validated 32-byte frame. Mirrors Python's `_parse(frame)`.
// Frame layout (big-endian): [6:8] PM2.5 standard, [8:10] PM10 standard.
func (c *Connector) parse(frame []byte) *connector.AirQualityData {
	word := func(offset int) int {
		return int(frame[offset])<<8 | int(frame[offset+1])
	}

	pm25 := float64(word(6))
	pm10 := float64(word(8))
	aqi := pm25ToAQI(pm25)

	data := connector.NewAirQualityData()
	data.AQI = &aqi
	data.PM25 = &pm25
	data.PM10 = &pm10
	data.Location = c.cfg.Location
	data.Source = "pms5003"
	return data
}

// pm25ToAQI mirrors Python's `_pm25_to_aqi(pm25)`: US EPA breakpoint formula.
func pm25ToAQI(pm25 float64) int {
	type bp struct{ cLow, cHigh, aqiLow, aqiHigh float64 }
	breakpoints := []bp{
		{0.0, 12.0, 0, 50},
		{12.1, 35.4, 51, 100},
		{35.5, 55.4, 101, 150},
		{55.5, 150.4, 151, 200},
		{150.5, 250.4, 201, 300},
		{250.5, 350.4, 301, 400},
		{350.5, 500.4, 401, 500},
	}
	for _, b := range breakpoints {
		if pm25 >= b.cLow && pm25 <= b.cHigh {
			aqi := ((b.aqiHigh-b.aqiLow)/(b.cHigh-b.cLow))*(pm25-b.cLow) + b.aqiLow
			return int(aqi + 0.5) // round half up, mirrors Python's round()
		}
	}
	return 500
}

